### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by dselans**
Some potential refactor work for dkenv. Will add tests shortly.

Don't merge - let's discuss first when we have a moment!

@mikkergimenez @idleyoungman

### CODE REVIEW

File: dkenv/dkenv.go [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+//     - backup destination file
+//     - create symlink
+// - If destination exists AND *is* a symlink:
+//     - check if the symlink already matches destination (noop if it does)
+//     - remove and overwrite if symlink is incorrect
+// - If destination does not exist:
+//     - create symlink
+//
+// TODO This will be a pain to test - refactor when possible.
+func (d *Dkenv) UpdateSymlink(version string) error {
```
Comment (dselans) [inline dkenv/dkenv.go]: This is really meh.. would like to update this prior to merge; it'll also make testing it _much_ simpler.

File: lib/downloader.go [hunk 2]
```diff
@@ -49,102 +31,81 @@ type PassThru struct {
...
-}
-
-func redirectPolicyFunc(req *http.Request, via []*http.Request) error {
-	if len(via) > 10 {
-		return Err_Too_Many_Redirect
-	}
-
-	if req.URL.Scheme != "https" {
-		return Err_HTTP_Redirect
+	if err := ioutil.WriteFile(d.DkenvDir+"/docker-"+version, body, 0777); err != nil {
```
Comment (idleyoungman) [inline lib/downloader.go]: What about 0755 for FileMode?

Comment (dselans) [inline lib/downloader.go]: Agreed - bad copy/paste.

**Comment by idleyoungman** [comment]
:joy_cat: LGTM!

**Comment by idleyoungman** [comment]
Fixed the merge conflicts while I was at it.

**Comment by dselans** [comment]
Added (some) tests, godeps - this should be in a good enough state at this point. Waiting on @mikkergimenez @idleyoungman for review.

Will flesh out the rest of the tests when I have a bit more time.

```
dselans:dkenv dselans$ godep-osx go test ./... -cover
?       github.com/newrelic/dkenv   [no test files]
ok      github.com/newrelic/dkenv/cli   0.013s  coverage: 89.1% of statements
ok      github.com/newrelic/dkenv/lib   0.015s  coverage: 0.9% of statements
```

**Pull request merged**